lib/IRGen/Outlining.cpp

```diff
@@ -67,18 +73,24 @@ void OutliningMetadataCollector::collectTypeMetadataForLayout(SILType type) {
   Values.insert({key, metadata});
 }
 
-void OutliningMetadataCollector::collectFormalTypeMetadata(CanType type) {
-  // If the type has no archetypes, we can emit it from scratch in the callee.
-  assert(type->hasArchetype());
+void OutliningMetadataCollector::collectTypeMetadataForEnvironment(SILType ty) {
+  auto bgt = dyn_cast<BoundGenericType>(ty.getASTType());
```

Comment:
This is insufficient:

- The BoundGenericType `Outer<Int>.Inner<String>` has a parent type `Outer<Int>` you must visit:
```
struct Outer<T> { struct Inner<U> { } }
```

- A NominalType might have a parent type:
```
struct Outer<T> { struct Inner { } }
```

Instead, you need to walk the replacement types in the context substitution map:

```
if (type->isSpecialized()) {
  auto subMap = type->getContextSubstitutionMap(module, type->getAnyNominal());
  for (auto replacementType : subMap.getReplacementTypes()) { ... }
}
```

Also, what about witness tables? The generic type might depend on a conformance:

```
struct Foo<T: P> {
  var x: T.A

  deinit {
    ... doing stuff with x requires T: P
  }
}
```

Reply:
Bonus: type in constrained extension depends on the constrained extension's conformance:

```
struct G<T> {}

extension G where T: Sequence {
  struct Inner: ~Copyable {
    var x: T.Element

    deinit { ... }
  }
}
```

Reply:
Yes, I am aware of these limitations.